### ISSUE TITLE
Need some help getting Docker up and running

### ISSUE THREAD

**Issue by bravo-kernel** [abchrisxyz/ergowatch#5]
Hi there, 

I am trying to start docker and would appreciate some help. 
It looks like some files related to the grabber are missing from the repository but I could be do something wrong as well.

> Removed original description after receiving your feedback

**Comment by abchrisxyz** [comment]
Hey,

ErgoWatch doesn't use ergo-boostrap.

There's a `build.sh` script in the repo root that'll prepare a Dockerfile for the grabber. That should solve the errors you ran into. Will add clearer instructions at some point...

If all you're after is getting the grabber running, have a look at https://www.christopheryee.org/blog/ergo-how-to-setup-blockchain-analytics-on-raspberry-pi/ not docker based, but details what needs to happen.

Good luck and come say hi in #ergo-stats on Ergo's Discord ;)

**Comment by bravo-kernel** [comment]
Ah, one of my famous "silly me" first issues :-D

Thanks for the pointers, I will give them a go tomorrow. Love your site, awesome job 💃

**Comment by bravo-kernel** [comment]
## Updated Installation Instructions

```bash
git clone https://github.com/abchrisxyz/ergowatch
cd ergowatch

# install grabber
./build.sh

# create Postgress database password file
touch ./explorer-backend/db/db.secret

# update Postgres database password file with your own password
POSTGRES_PASSWORD=hello123

## Build and run the docker environment
docker-compose -f docker-compose.yml up -d --build
```

## Error

```
 => [ergowatch_grabber builder 4/6] WORKDIR /explorer-backend                                                                                                                                      0.0s 
 => [ergowatch_grabber builder 5/6] RUN sbt chain-grabber/assembly                                                                                                                               194.5s 
 => [ergowatch_grabber builder 6/6] RUN mv `find . -name ChainGrabber-assembly-*.jar` /chain-grabber.jar                                                                                           1.1s 
 => [ergowatch_grabber stage-1 2/2] COPY --from=builder /chain-grabber.jar /chain-grabber.jar                                                                                                      0.3s 
[+] Running 1/1
 - Network ergowatch_default  Created                                                                                                                                                              0.0s 

external volume "" not found
The terminal process "docker-compose -f "docker-compose.yml" up -d --build" terminated with exit code: 1.
```

**Comment by bravo-kernel** [comment]
## Workaround?

Unsure why but changing the volumes `external` boolean in `docker-compose.yml` to `false`, like shown below, solves the issue for me (using Docker Desktop on Windows).

```yml
volumes:
  ew_node:
    external: false
```

All containers are now started as expected 💃 :

```bash
 - Container ergowatch_node_1        Running       0.0s
 - Container ergowatch_db_1          Running       0.0s
 - Container ergowatch_grabber_1     Running       0.0s
 - Container ergowatch_syncer_1      Started       0.1s
 - Container ergowatch_api_1         Started       0.3s
```

**Comment by abchrisxyz** [comment]
Well done. Yes, I use external volumes to avoid dropping them accidentally. So you have to create it manually before starting the containers: `docker volume create ew_node`

**Comment by abchrisxyz** [comment]
Note that you may run into some slowdown trying to sync the grabber from scratch. You can speed things up by dropping the indexes in the grabber db schema, sync it to current height, then recreate the indexes.

**Comment by bravo-kernel** [comment]
Thank you kindly sir 💯 

One more issue to resolve. It looks like the following containers are not able to make the database connection:

- ergowatch_db_1
- ergowatch_api_1
- ergowatch_syncer_1

FWIW I know that the env file is being used because I am seeing `hello123` in the logs which is my `POSTGRES_PASSWORD`. 

### ergowatch_api_1 (and ergowatch_syncer_1)

Error:

```
asyncpg.exceptions.InvalidPasswordError: password authentication failed for user "hello123"
```

<details>
  <summary>Log dump</summary>

```bash
[2021-12-31 09:42:11 +0000] [9] [ERROR] Application startup failed. Exiting.

[2021-12-31 09:42:11 +0000] [9] [INFO] Worker exiting (pid: 9)

[2021-12-31 09:42:11 +0000] [24] [INFO] Booting worker with pid: 24

[2021-12-31 09:42:11 +0000] [11] [INFO] Started server process [11]

[2021-12-31 09:42:11 +0000] [11] [INFO] Waiting for application startup.

{"loglevel": "info", "workers": 8, "bind": "0.0.0.0:80", "graceful_timeout": 120, "timeout": 120, "keepalive": 5, "errorlog": "-", "accesslog": "-", "workers_per_core": 1.0, "use_max_workers": null, "host": "0.0.0.0", "port": "80"}

{"loglevel": "info", "workers": 8, "bind": "0.0.0.0:80", "graceful_timeout": 120, "timeout": 120, "keepalive": 5, "errorlog": "-", "accesslog": "-", "workers_per_core": 1.0, "use_max_workers": null, "host": "0.0.0.0", "port": "80"}

{"loglevel": "info", "workers": 8, "bind": "0.0.0.0:80", "graceful_timeout": 120, "timeout": 120, "keepalive": 5, "errorlog": "-", "accesslog": "-", "workers_per_core": 1.0, "use_max_workers": null, "host": "0.0.0.0", "port": "80"}

{"loglevel": "info", "workers": 8, "bind": "0.0.0.0:80", "graceful_timeout": 120, "timeout": 120, "keepalive": 5, "errorlog": "-", "accesslog": "-", "workers_per_core": 1.0, "use_max_workers": null, "host": "0.0.0.0", "port": "80"}

Checking for script in /app/prestart.sh

There is no script /app/prestart.sh

[2021-12-31 09:42:11 +0000] [10] [ERROR] Traceback (most recent call last):

  File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 540, in lifespan

    async for item in self.lifespan_context(app):

  File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 481, in default_lifespan

    await self.startup()

  File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 516, in startup

    await handler()

  File "/app/main.py", line 28, in startup_event

    await db.init_connection_pool()

  File "/app/db.py", line 12, in init_connection_pool

    CONNECTION_POOL = await asyncpg.create_pool(dbstr)

  File "/usr/local/lib/python3.8/site-packages/asyncpg/pool.py", line 413, in _async__init__

    await self._initialize()

  File "/usr/local/lib/python3.8/site-packages/asyncpg/pool.py", line 441, in _initialize

    await first_ch.connect()

  File "/usr/local/lib/python3.8/site-packages/asyncpg/pool.py", line 133, in connect

    self._con = await self._pool._get_new_connection()

  File "/usr/local/lib/python3.8/site-packages/asyncpg/pool.py", line 511, in _get_new_connection

    con = await connection.connect(

  File "/usr/local/lib/python3.8/site-packages/asyncpg/connection.py", line 2085, in connect

    return await connect_utils._connect(

  File "/usr/local/lib/python3.8/site-packages/asyncpg/connect_utils.py", line 881, in _connect

    return await _connect_addr(

  File "/usr/local/lib/python3.8/site-packages/asyncpg/connect_utils.py", line 781, in _connect_addr

    return await __connect_addr(params, timeout, True, *args)

  File "/usr/local/lib/python3.8/site-packages/asyncpg/connect_utils.py", line 831, in __connect_addr

    await compat.wait_for(connected, timeout=timeout)

  File "/usr/local/lib/python3.8/site-packages/asyncpg/compat.py", line 66, in wait_for

    return await asyncio.wait_for(fut, timeout)

  File "/usr/local/lib/python3.8/asyncio/tasks.py", line 494, in wait_for

    return fut.result()

asyncpg.exceptions.InvalidPasswordError: password authentication failed for user "hello123"
```
</details>


### ergowatch_db_1

Different error message.

<details>
  <summary>Log dump</summary>

```bash
2021-12-31 10:02:23.892 UTC [284] FATAL:  password authentication failed for user "hello123"

2021-12-31 10:02:23.892 UTC [284] DETAIL:  Role "hello123" does not exist.

Connection matched pg_hba.conf line 100: "host all all all scram-sha-256"

2021-12-31 10:02:33.662 UTC [286] FATAL:  password authentication failed for user "ergo"

2021-12-31 10:02:33.662 UTC [286] DETAIL:  Connection matched pg_hba.conf line 100: "host all all all scram-sha-256"
```
</details>

**Comment by abchrisxyz** [comment]
The db user is defined the postgres Dockerfile, change it to whatever you want, or use the `ergo` user that is defined currently (you can still use whatever password you set in `./explorer-backend/db/db.secret`).

**Comment by bravo-kernel** [comment]
Right but... I am doing exactly that so user is `ergo` and password is coming from my `db.secret`. 

However, mentioned containers throw errors making the connection. We could do a Discord to discuss further if need be. Feel free to DM me.

**Comment by bravo-kernel** [comment]
To reproduce:

```bash
git clone https://github.com/abchrisxyz/ergowatch
cd ergowatch

# install grabber
./build.sh

# create Postgres database password file
touch ./explorer-backend/db/db.secret

#  update newly created Postgress database password file
POSTGRES_PASSWORD=your-database-password

# Create external docker volume for ergowatch
docker volume create ew_node

# Build and run the docker environement
docker-compose -f docker-compose.yml up -d --build
```

Next check container statuses, restarting because of failed database connection:

![image](https://user-images.githubusercontent.com/230500/147875406-7763af38-92dc-4ad1-a73c-a4c497ca1f21.png)

**Comment by abchrisxyz** [comment]
Found ya on Discord, but you're not accepting friends.

What error message do you get from the syncer and api containers now?
Try deleting/recreating the db volume to ensure the db is initialized from scratch.

**Comment by bravo-kernel** [comment]
Ah, I see. Seems you have the same friends-limitation but just sent you an invite.

**Comment by bravo-kernel** [comment]
All resolved in #8
